### PR TITLE
Add timer interface with millisecond resolution to python plugin

### PR DESCRIPTION
**Pull request by xrmx**
We extend uwsgi timer api to handle a nanosecond value in order to support timer with finer grained resolution than the current second. Then we add interface to use timers with millisecond resolution is added to the python plugin. I've extended the decorators with another decorator mstimer. Alternatively we can add a resolution parameter to the current timer decorator. It's fine for me either way.
Eventually external plugins that used that interface of course need to be updated.
Please note this has NOT ever been compiled on kqueue systems.

### CODE REVIEW

File: core/event.c [hunk 1]
```diff
@@ -1382,7 +1382,7 @@ int event_queue_add_timer(int eq, int *id, int sec) {
 	*id = timer_id;
 	timer_id++;
 
-	EV_SET(&kev, *id, EVFILT_TIMER, EV_ADD, 0, sec * 1000, 0);
+	EV_SET(&kev, *id, EVFILT_TIMER, EV_ADD, 0, sec * 1000 + int(nsec / 1000000), 0);
```
Comment (akx) [inline core/event.c]: The unit of the 6th parameter isn't immediately clear... Could it somehow be made clearer?

Comment (xrmx) [inline core/event.c]: Sure, we may use and intermediate variable to make it explicit.

**Comment by Darvame** [comment]
Hi, lua plugin also uses <b>uwsgi_add_timer()</b>
and, in struct uwsgi_timer, order like:
```c
int registered;
uint8_t sig;
\\7 bytes gap?
long nsvalue;
```
is not cool for structure packing (struct memory usage)
but this is not so important

**Comment by xrmx** [comment]
@Darvame i know but we tend to not breaking ABI and add new struct members at the end. Now i don't know if that struct is an internal one or not. We should probably start thinking to separate the public stuff for the internal one in uwsgi.h.

**Comment by xrmx** [comment]
from unbit:
```
i would prefer to not break the old api, what about defining event_queue_add_timer_hr and leaving event_queue_add_timer as is ?
```

**Comment by xrmx** [comment]
Updated pull requests, anyone that can give it a try on a freebsd / mac os x machine appreciated :) cc @unbit

**Comment by Darvame** [comment]
um, 
```bash
core/event.c:1385:39: error: expected expression
        const long timeout_ms = sec * 1000 + int(nsec / 1000000);
```
on my mac with clang

**Comment by xrmx** [comment]
@Darvame thanks for checking. Writing a lot more pytohn than C these days. Pushed a new version and another patch.

**Comment by Darvame** [comment]
```bash
core/event.c:1384:10: error: read-only variable is not assignable
        timer_id++;
```
timer_id must be converted from const before ++?

**Comment by xrmx** [comment]
@Darvame thanks, the last patch was completely non sense, removed.

**Comment by Darvame** [comment]
clang version
```bash
Apple LLVM version 7.0.0 (clang-700.1.76)
Target: x86_64-apple-darwin15.0.0
Thread model: posix
```
python3.4

```bash
git clone https://github.com/xrmx/uwsgi.git -b nsectimer2
cd uwsgi
export CC=clang
python3 uwsgiconfig.py --build
./uwsgi --wsgi-file t/python/timers.py --http :8080 --master
```
no problems so far

**Comment by xrmx** [comment]
@Darvame Thanks for testing!

**Comment by unbit** [comment]
Thanks to all

**Pull request merged**